src/semaphore.ts

```diff
@@ -0,0 +1,46 @@
+type Unlock = () => void;
+
+export class Semaphore {
+
+  private _maxAllowed: number;
+  private _permits: number;
+  private _queue: Unlock[] = [];
+
+  constructor(maxAllowed: number = 1) {
+    this._maxAllowed = maxAllowed;
+    this._permits = maxAllowed;
+  }
+
+  get permits() {
+    return this._permits;
+  }
+
+  get queueLength() {
+    return this._queue.length;
+  }
+
+  get acquired() {
+    return this._maxAllowed - this._permits;
+  }
+
+  public async acquire(): Promise<void> {
+    if (this._permits > 0) {
+      this._permits -= 1;
+      return;
+    }
+
+    return new Promise(resolve => this._queue.push(resolve));
+  }
+
+  public release(): void {
+    if (this._queue.length) {
+      this._queue.shift()?.();
+    } else {
+      this._permits += 1;
+    }
+
+    if (this._permits > this._maxAllowed) {
+      console.error('Release() has been called without previous acquire!');
+    }
+  }
+}
```

Comment:
The code is for a Semaphore class written in TypeScript, designed to control concurrent access to shared resources by multiple processes. Here's a brief review with some potential issues and suggestions:

1. **Thread Safety**: Since JavaScript runs in a single-threaded environment (apart from Web Workers), this semaphore implementation should generally work fine without concerns about traditional thread safety. However, it relies on the event loop to queue and execute async operations correctly; be aware of possible issues if used in environments beyond browser or Node.js.

2. **Error Handling**: There are no checks if `maxAllowed` provided in the constructor is non-negative or not zero. Also, releasing permits more times than they were acquired should have better error handling, potentially throwing an Error instead of just logging to the console. This might allow clients to catch the mistake and respond accordingly.

3. **Race Condition on Release**: A situation where `release` is called while the permit count `_permits` equals `_maxAllowed` isn't handled. A check to make sure `_permits <= _maxAllowed` before incrementing `_permits` would avoid over-releasing.

4. **Asynchronous Design**: The `acquire` method allows you to wait asynchronously until you can acquire a permit. It correctly queues up callers when there are no permits available. Make sure calling contexts correctly handle the Promise returned from `acquire`, possibly causing an unhandled rejection if not managed well, especially if `acquire()` is modified to throw errors in the future.

5. **Queue Implementation**: Using `Array.shift()` can be inefficient for large queues as it has O(n) time complexity because it needs to reindex the array after removal. If performance with large numbers of queued acquires is a concern, consider using a more efficient queue data structure.

6. **Semaphore Release Logic**: As of now, `release` prioritizes dequeuing waiting consumers over incrementing `_permits`. This behavior should be clearly documented as it affects how the semaphore works.

7. **Overflow Checking**: No checks for integer overflows on `_permits`. JavaScript numbers can safely represent integers up to 2^53 - 1, but misuse or extreme usage could still result in problems.

8. **Code Cleanup**: Some semaphores include a method to forcibly remove all waiting locks. Consider providing such a reset functionality for cases where recovery from a locked state is required.

9. **Informative Errors**: When errors do occur, such as over releasing, provide more informative error messages or custom Error types that can help users identify and debug issues.

10. **Documentation and Typing**: While type annotations are present, explicit return types on functions could be added for clarity. Additionally, comprehensive documentation in terms of JSDoc or comments detailing the methods' behaviors, edge cases, and expected usage patterns would be immensely helpful for other developers.

11. **Testing**: Ensure that this code is well tested, particularly around edge cases like releasing too many times, concurrent acquires, and proper increment and decrement of permits.

In practice, modify and improve upon this class depending on the actual concurrency model in which it will be used.